sys/net/gnrc/link_layer/gomach/gomach.c

```diff
@@ -0,0 +1,2193 @@
+/*
+ * Copyright (C) 2016 INRIA
+ *
+ * This file is subject to the terms and conditions of the GNU Lesser
+ * General Public License v2.1. See the file LICENSE in the top level
+ * directory for more details.
+ */
+
+/**
+ * @ingroup     net_gnrc_gomach
+ * @{
+ *
+ * @file
+ * @brief       Implementation of GoMacH
+ *
+ * @author      Shuguo Zhuo  <shuguo.zhuo@inria.fr>
+ * @}
+ */
+
+#include <errno.h>
+#include <stdlib.h>
+#include <string.h>
+#include <stdint.h>
+#include <stdbool.h>
+
+#include "random.h"
+#include "timex.h"
+#include "periph/rtt.h"
+#include "net/gnrc/netif.h"
+#include "net/gnrc/netif/internal.h"
+#include "net/gnrc/netif/ieee802154.h"
+#include "net/netdev/ieee802154.h"
+#include "net/gnrc.h"
+#include "net/gnrc/nettype.h"
+#include "net/netdev.h"
+#include "net/gnrc/mac/internal.h"
+#include "net/gnrc/gomach/gomach.h"
+#include "net/gnrc/gomach/timeout.h"
+#include "include/gomach_internal.h"
+
+#define ENABLE_DEBUG    (0)
+#include "debug.h"
+
+#ifndef LOG_LEVEL
+/**
+ * @brief Default log level define
+ */
+#define LOG_LEVEL LOG_WARNING
+#endif
+
+#include "log.h"
+
+/**
+ * @brief  GoMacH thread's PID
+ */
+static kernel_pid_t gomach_pid;
+
+static void _gomach_init(gnrc_netif_t *netif);
+static int _send(gnrc_netif_t *netif, gnrc_pktsnip_t *pkt);
+static gnrc_pktsnip_t *_recv(gnrc_netif_t *netif);
+static void _gomach_msg_handler(gnrc_netif_t *netif, msg_t *msg);
+static void _gomach_event_cb(netdev_t *dev, netdev_event_t event);
+
+static const gnrc_netif_ops_t gomach_ops = {
+    .init = _gomach_init,
+    .send = _send,
+    .recv = _recv,
+    .get = gnrc_netif_get_from_netdev,
+    .set = gnrc_netif_set_from_netdev,
+    .msg_handler = _gomach_msg_handler,
+};
+
+gnrc_netif_t *gnrc_netif_gomach_create(char *stack, int stacksize,
+                                       char priority, char *name,
+                                       netdev_t *dev)
+{
+    return gnrc_netif_create(stack, stacksize, priority, name, dev,
+                             &gomach_ops);
+}
+
+static gnrc_pktsnip_t *_recv(gnrc_netif_t *netif)
+{
+    netdev_t *dev = netif->dev;
+    netdev_ieee802154_rx_info_t rx_info;
+    netdev_ieee802154_t *state = (netdev_ieee802154_t *)netif->dev;
+    gnrc_pktsnip_t *pkt = NULL;
+    int bytes_expected = dev->driver->recv(dev, NULL, 0, NULL);
+
+    if (bytes_expected > 0) {
+        int nread;
+
+        pkt = gnrc_pktbuf_add(NULL, NULL, bytes_expected, GNRC_NETTYPE_UNDEF);
+        if (pkt == NULL) {
+            DEBUG("_recv_ieee802154: cannot allocate pktsnip.\n");
+            return NULL;
+        }
+        nread = dev->driver->recv(dev, pkt->data, bytes_expected, &rx_info);
+        if (nread <= 0) {
+            gnrc_pktbuf_release(pkt);
+            return NULL;
+        }
+        if (!(state->flags & NETDEV_IEEE802154_RAW)) {
+            nread = dev->driver->recv(dev, pkt->data, bytes_expected, &rx_info);
```

Comment:
Also, where did the header parsing to `netif_hdr` go now? Can't find it....

Reply:
Oops, I really don't get it, sorry about this!
So, just to make it clear and settle this problem totally, is it the expected `recv` func like this (no `gnrc_netif_hdr_t` header anymore):
```
static gnrc_pktsnip_t *_recv(gnrc_netif_t *netif)
{
    netdev_t *dev = netif->dev;
    netdev_ieee802154_rx_info_t rx_info;
    netdev_ieee802154_t *state = (netdev_ieee802154_t *)netif->dev;
    gnrc_pktsnip_t *pkt = NULL;
    int bytes_expected = dev->driver->recv(dev, NULL, 0, NULL);

    if (bytes_expected > 0) {
        int nread;

        pkt = gnrc_pktbuf_add(NULL, NULL, bytes_expected, GNRC_NETTYPE_UNDEF);
        if (pkt == NULL) {
            DEBUG("_recv_ieee802154: cannot allocate pktsnip.\n");
            return NULL;
        }
        nread = dev->driver->recv(dev, pkt->data, bytes_expected, &rx_info);
        if (nread <= 0) {
            gnrc_pktbuf_release(pkt);
            return NULL;
        }
        if (!(state->flags & NETDEV_IEEE802154_RAW)) {
            gnrc_pktsnip_t *ieee802154_hdr, *netif_hdr;
#if ENABLE_DEBUG
            char src_str[GNRC_NETIF_HDR_L2ADDR_PRINT_LEN];
#endif
            size_t mhr_len = ieee802154_get_frame_hdr_len(pkt->data);

            if (mhr_len == 0) {
                DEBUG("_recv_ieee802154: illegally formatted frame received\n");
                gnrc_pktbuf_release(pkt);
                return NULL;
            }
            nread -= mhr_len;
            /* mark IEEE 802.15.4 header */
            ieee802154_hdr = gnrc_pktbuf_mark(pkt, mhr_len, GNRC_NETTYPE_UNDEF);
            if (ieee802154_hdr == NULL) {
                DEBUG("_recv_ieee802154: no space left in packet buffer\n");
                gnrc_pktbuf_release(pkt);
                return NULL;
            }
            netif_hdr = _make_netif_hdr(ieee802154_hdr->data);
            if (netif_hdr == NULL) {
                DEBUG("_recv_ieee802154: no space left in packet buffer\n");
                gnrc_pktbuf_release(pkt);
                return NULL;
            }

            pkt->type = state->proto;
#if ENABLE_DEBUG
            DEBUG("_recv_ieee802154: received packet from %s of length %u\n",
                  gnrc_netif_addr_to_str(src_str, sizeof(src_str),
                                         gnrc_netif_hdr_get_src_addr(hdr),
                                         hdr->src_l2addr_len),
                  nread);
#if defined(MODULE_OD)
            od_hex_dump(pkt->data, nread, OD_WIDTH_DEFAULT);
#endif
#endif
            gnrc_pktbuf_remove_snip(pkt, ieee802154_hdr);
            LL_APPEND(pkt, netif_hdr);
        }

        DEBUG("_recv_ieee802154: reallocating.\n");
        gnrc_pktbuf_realloc_data(pkt, nread);
    }

    return pkt;
}
```

Reply:
> (no gnrc_netif_hdr_t header anymore):

You *are* creating the header here...

```C
netif_hdr = _make_netif_hdr(ieee802154_hdr->data);
if (netif_hdr == NULL) {
    DEBUG("_recv_ieee802154: no space left in packet buffer\n");
    gnrc_pktbuf_release(pkt);
    return NULL;
}
```

Ok then... Let's take ten steps back and start from the beginning: I think [this extension](https://github.com/RIOT-OS/RIOT/pull/5618/files#diff-05805abfe6eb0ea11585cea5ebd029b6R91) of the `gnrc_netif_hdr_t` struct is not necessary and I really don't want it to be there unless it is absolutely necessary, since it is a pretty central struct to GNRC. Another reason is, that the sequence number is pretty link-layer specific and the *generic* network interface header (aka `gnrc_netif_hdr_t` ;-)) is supposed to be independent of link-layer specific information (WiFi e.g. uses 12, instead of 8, bits for its sequence number and that doesn't even count the additional *fragment* number).

Here is why: The only code fragment that needs this extension is this MAC protocol. However, all context it requires (specifically the sequence number) for its operation should as far as I can see it be readable from the original MAC/L2 header. So why do you discard it here (and replace it with an extended header `gnrc_netif_hdr_t`)? I proposed the following:

* don't discard the original IEEE 802.15.4 header and *do not* create a `gnrc_netif_hdr_t` from its information here in the receive function. *At maximum* mark the IEEE 802.15.4 (you do that already, but you replace it with the `gnrc_netif_hdr_t`), this way you still can read the sequence number using `ieee802154_get_seq(pkt->next->data)` (remember, that `ieee802154` header is only valid in this frame):

```C
gnrc_pktsnip_t *ieee802154_hdr;
size_t mhr_len = ieee802154_get_frame_hdr_len(pkt->data);

if (mhr_len == 0) {
    DEBUG("_recv_ieee802154: illegally formatted frame received\n");
    gnrc_pktbuf_release(pkt);
    return NULL;
}
nread -= mhr_len;
/* mark IEEE 802.15.4 header */
ieee802154_hdr = gnrc_pktbuf_mark(pkt, mhr_len, GNRC_NETTYPE_UNDEF);
if (ieee802154_hdr == NULL) {
    DEBUG("_recv_ieee802154: no space left in packet buffer\n");
    gnrc_pktbuf_release(pkt);
    return NULL;
}
```

(you also can still leave the `l2filter`-functionality in here, by using `ieee802154_get_src()` instead of what you were doing with the already parsed information)

* I'm not deep enough into your code to say this with confidence (so from here on out also don't just copy what I write, but think about it, if it makes sense) but I assume after calling this receive function [here](https://github.com/RIOT-OS/RIOT/pull/5618/files#diff-4b2a0337b422754c37b9ad5000ed9c30R2030), you put it in some kind of receive queue (you don't dispatch it directly to the next layer at least using `gnrc_netapi_dispatch_receive()`).
* At some point ([here?](https://github.com/RIOT-OS/RIOT/pull/5618/files#diff-d4ab863beb293e5d6f5ce98af945e701R268) however you *are* dispatching the header. This is the point where other layers of GNRC require the generic interface header and GoMacH is seizing control over it (i.e. the sequence number isn't needed anymore as well). As such it is here where you *have to* create the `gnrc_netif_hdr_t` at the latest (maybe instead of [this line](https://github.com/RIOT-OS/RIOT/pull/5618/files#diff-d4ab863beb293e5d6f5ce98af945e701R258), or maybe even better before calling `gnrc_mac_dispatch()`, so this behavior stays in GoMacH? Following code is based on the assumption):

```C
void gnrc_gomach_dispatch(gnrc_netif_t *netif) {
    gnrc_mac_rx_t *rx = &netif->mac.rx;
    netdev_ieee802154_t *state = (netdev_ieee802154_t *)netif->dev;
    gnrc_pktsnip_t *netif = _make_netif_hdr(rx->dispatch_buffer[i]->next->next->data);

    if (netif == NULL) {
        DEBUG("gnrc_gomach_dispatch: no space left in packet buffer\n");
        gnrc_pktbuf_release(pkt);
        return NULL;
    }
    gnrc_netif_hdr_t *hdr = netif_hdr->data;
    /* Maybe you can queue rx_info you received from `netdev->driver->recv()` in `netif->ops->recv()`
     * in gnrc_gomach_t as well? */
    /* hdr->lqi = rx_info.lqi; */
    /* hdr->rssi = rx_info.rssi; */
    hdr->if_pid = thread_getpid();
    rx->dispatch_buffer[i]->next->type = state->proto;
    gnrc_pktbuf_remove_snip(rx->dispatch_buffer[i]->next, rx->dispatch_buffer[i]->next->next);
    gnrc_mac_dispatch(rx);
}
```

I think this would be an elegant solution to avoid extending `gnrc_netif_hdr_t` just for this protocol.

I hope you can get the relevant information from this lengthy comment. If you have questions: just ask.

Reply:
Thanks a lot for the very detailed description! Working on it!~ Hope I do right this time~